File: solutions/prospective/algorithms/patterns/lua/bfs_queue_grid.lua
--[[
BFS on Grid (Queue) Template

Shortest path in unweighted grid.
- Use table as queue (table.remove(t, 1) is O(n), for O(1) use two-pointer index or linked list).
- Track visited to avoid cycles.
--]]
-- === MEMO START ===
function bfs_grid(grid, start_r, start_c)
    local rows = #grid
    local cols = #grid[1]
    local q = {{r = start_r, c = start_c, dist = 0}}
    local visited = {}
    visited[start_r .. "," .. start_c] = true
    
    local directions = {{0, 1}, {0, -1}, {1, 0}, {-1, 0}}
    
    while #q > 0 do
        local current = table.remove(q, 1)
        local r, c, dist = current.r, current.c, current.dist
        
        -- if r == target_r and c == target_c then return dist end
        
        for _, dir in ipairs(directions) do
            local nr = r + dir[1]
            local nc = c + dir[2]
            
            if nr >= 1 and nr <= rows and nc >= 1 and nc <= cols then
                local key = nr .. "," .. nc
                if not visited[key] and grid[nr][nc] ~= '#' then
                    visited[key] = true
                    table.insert(q, {r = nr, c = nc, dist = dist + 1})
                end
            end
        end
    end
    
    return -1
end

File: solutions/prospective/algorithms/patterns/c/bfs_queue_grid.c
/*
BFS on Grid (Queue) Template

Shortest path in unweighted grid.
- Use a Queue for FIFO.
- Track visited to avoid cycles.
*/
#include <stdlib.h>
#include <stdbool.h>

// === MEMO START ===
typedef struct {
    int r;
    int c;
    int dist;
} Node;

typedef struct {
    Node *data;
    int front;
    int rear;
    int capacity;
} Queue;

Queue* createQueue(int capacity) {
    Queue *q = (Queue *)malloc(sizeof(Queue));
    q->capacity = capacity;
    q->front = 0;
    q->rear = 0;
    q->data = (Node *)malloc(capacity * sizeof(Node));
    return q;
}

void enqueue(Queue *q, int r, int c, int dist) {
    if (q->rear == q->capacity) return; // Full
    q->data[q->rear].r = r;
    q->data[q->rear].c = c;
    q->data[q->rear].dist = dist;
    q->rear++;
}

Node dequeue(Queue *q) {
    return q->data[q->front++];
}

bool isEmpty(Queue *q) {
    return q->front == q->rear;
}

int bfs_grid(char **grid, int rows, int cols, int start_r, int start_c) {
    Queue *q = createQueue(rows * cols);
    bool **visited = (bool **)malloc(rows * sizeof(bool *));
    for (int i = 0; i < rows; i++) {
        visited[i] = (bool *)calloc(cols, sizeof(bool));
    }

    enqueue(q, start_r, start_c, 0);
    visited[start_r][start_c] = true;

    int dr[] = {0, 0, 1, -1};
    int dc[] = {1, -1, 0, 0};

    while (!isEmpty(q)) {
        Node current = dequeue(q);
        int r = current.r;
        int c = current.c;
        int dist = current.dist;

        // if (r, c) is target: return dist;
        
        for (int i = 0; i < 4; i++) {
            int nr = r + dr[i];
            int nc = c + dc[i];

            if (nr >= 0 && nr < rows && nc >= 0 && nc < cols &&
                !visited[nr][nc] && grid[nr][nc] != '#') {
                
                visited[nr][nc] = true;
                enqueue(q, nr, nc, dist + 1);
            }
        }
    }

    // Cleanup
    for (int i = 0; i < rows; i++) free(visited[i]);
    free(visited);
    free(q->data);
    free(q);

    return -1;
}

